File: java-samples/Features/state-reissuance/workflows/src/main/java/net/corda/samples/statereissuance/flows/ExitLandTitleFlow.java
package net.corda.samples.statereissuance.flows;

import co.paralleluniverse.fibers.Suspendable;
import net.corda.core.contracts.StateAndRef;
import net.corda.core.contracts.StateRef;
import net.corda.core.flows.*;
import net.corda.core.transactions.SignedTransaction;
import net.corda.core.transactions.TransactionBuilder;
import net.corda.samples.statereissuance.contracts.LandTitleContract;
import net.corda.samples.statereissuance.states.LandTitleState;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class ExitLandTitleFlow {

    @InitiatingFlow
    @StartableByRPC
    public static class Initiator extends FlowLogic<SignedTransaction> {

        private StateRef stateRef;

        public Initiator(StateRef stateRef) {
            this.stateRef = stateRef;
        }

        @Override
        @Suspendable
        public SignedTransaction call() throws FlowException {
            List<StateAndRef<LandTitleState>> landTitleStateAndRefs = getServiceHub().getVaultService()
                    .queryBy(LandTitleState.class).getStates();

            StateAndRef<LandTitleState> inputStateAndRef = landTitleStateAndRefs.stream()
                    .filter(landTitleStateStateAndRef -> landTitleStateStateAndRef.getRef().equals(stateRef))
                    .findAny().orElseThrow(() -> new IllegalArgumentException("Land Title Not Found"));

            LandTitleState input = inputStateAndRef.getState().getData();

            TransactionBuilder builder = new TransactionBuilder(inputStateAndRef.getState().getNotary())
                    .addInputState(inputStateAndRef)
                    .addCommand(new LandTitleContract.Commands.Exit(), Arrays.asList(getOurIdentity().getOwningKey(),
                            input.getIssuer().getOwningKey()));

            builder.verify(getServiceHub());
            final SignedTransaction ptx = getServiceHub().signInitialTransaction(builder);

            FlowSession issuerSession = initiateFlow(input.getIssuer());

            SignedTransaction stx = subFlow(new CollectSignaturesFlow(ptx, Collections.singletonList(issuerSession)));
            return subFlow(new FinalityFlow(stx, Collections.singletonList(issuerSession)));
        }
    }

    @InitiatedBy(Initiator.class)
    public static class Responder extends FlowLogic<Void>{

        private FlowSession counterpartySession;

        public Responder(FlowSession counterpartySession) {
            this.counterpartySession = counterpartySession;
        }

        @Override
        @Suspendable
        public Void call() throws FlowException {

            subFlow(new SignTransactionFlow(counterpartySession) {
                @Suspendable
                @Override
                protected void checkTransaction(SignedTransaction stx) throws FlowException {

                }
            });

            //Stored the transaction into data base.
            subFlow(new ReceiveFinalityFlow(counterpartySession));
            return null;
        }
    }
}

File: kotlin-samples/Features/state-reissuance/workflows/src/main/kotlin/net/corda/samples/statereissuance/flows/ExitLandTitleFlow.kt
package net.corda.samples.statereissuance.flows

import co.paralleluniverse.fibers.Suspendable
import net.corda.core.contracts.StateRef
import net.corda.core.contracts.requireThat
import net.corda.core.flows.*
import net.corda.core.transactions.SignedTransaction
import net.corda.core.transactions.TransactionBuilder
import net.corda.core.utilities.ProgressTracker
import net.corda.samples.statereissuance.contracts.LandTitleContract
import net.corda.samples.statereissuance.states.LandTitleState

@InitiatingFlow
@StartableByRPC
class ExitLandTitle(private val stateRef: StateRef) : FlowLogic<SignedTransaction>() {
    override val progressTracker = ProgressTracker()

    @Suspendable
    override fun call(): SignedTransaction {

        val landTitleStateAndRefs = serviceHub.vaultService.queryBy(LandTitleState::class.java).states

        val inputStateAndRef = landTitleStateAndRefs.stream().filter{ it.ref == stateRef }
                .findAny().orElseThrow{IllegalArgumentException("Land Title Not Found")}

        val inputState = inputStateAndRef.state.data
        val notary = inputStateAndRef.state.notary

        val builder = TransactionBuilder(notary)
                .addInputState(inputStateAndRef)
                .addCommand(LandTitleContract.Commands.Exit(), listOf(inputState.issuer.owningKey, ourIdentity.owningKey))

        builder.verify(serviceHub)
        val ptx = serviceHub.signInitialTransaction(builder)

        val issuerSession = initiateFlow(inputState.issuer)

        val stx = subFlow(CollectSignaturesFlow(ptx, listOf(issuerSession)))
        return subFlow(FinalityFlow(stx, listOf(issuerSession)))
    }
}

@InitiatedBy(ExitLandTitle::class)
class ExitLandTitleResponder(val counterpartySession: FlowSession) : FlowLogic<SignedTransaction>() {
    @Suspendable
    override fun call(): SignedTransaction {
        val signTransactionFlow = object : SignTransactionFlow(counterpartySession) {
            override fun checkTransaction(stx: SignedTransaction) = requireThat {
                //Addition checks
            }
        }
        val txId = subFlow(signTransactionFlow).id
        return subFlow(ReceiveFinalityFlow(counterpartySession, expectedTxId = txId))
    }
}
